src/commerce-handler.js

```diff
@@ -1,74 +1,183 @@
+                            userId = userIdentities["other4"];
+                            break;
+                        case 'deviceApplicationStamp':
+                            userId = window.mParticle.getDeviceId();
+                            break;
+                        default:
+                            userId = null;
+                    }
+
+                    if (!userId) {
+                        userId = window.mParticle.getDeviceId();
+                    }
+                }
+
+                if (
+                    event.EventCategory ===
+                        mParticle.CommerceEventType.ProductPurchase ||
+                    event.EventCategory === mParticle.CommerceEventType.ProductRefund
+                ) {
+                    if (expandedEvent.EventName.indexOf('Total') > -1) {
+                        if (
+                            event.CustomFlags &&
+                            event.CustomFlags['OptimizelyFullStack.EventName']
+                        ) {
+                            eventKey =
+                                event.CustomFlags['OptimizelyFullStack.EventName'];
+                        } else {
+                            eventKey = expandedEvent.EventName;
+                        }
+
+                        // Overall purchase event
+                        if (
+                            expandedEvent.EventAttributes &&
+                            expandedEvent.EventAttributes['Total Amount']
+                        ) {
+                            eventTags.revenue =
+                                expandedEvent.EventAttributes['Total Amount'] * 100;
+                        }
+                        // other individual product events should not have revenue tags
+                        // which are added via the expandCommerceEvent method above
+                    } else {
+                        if (
+                            event.CustomFlags &&
+                            event.CustomFlags['OptimizelyFullStack.EventName']
+                        ) {
+                            eventKey =
+                                event.CustomFlags['OptimizelyFullStack.EventName'];
+                        }
+                        if (eventTags.revenue) {
+                            delete eventTags.revenue;
+                        }
+                    }
+                } else {
+                    eventKey = expandedEvent.EventName;
+                    if (
+                        event.CustomFlags &&
+                        event.CustomFlags['OptimizelyFullStack.EventName']
+                    ) {
+                        eventKey =
+                            event.CustomFlags['OptimizelyFullStack.EventName'];
+                    }
+                }
             }
-            window['optimizely'].push(eventCopy);
+
+            window['optimizelyClientInstance'].track(eventKey, userId, userAttributes, eventTags);
```

Comment:
this should move up 2 lines to be in the `if` statement that ends in line 176.

add the following test. 

```
it.only('should not track a commerce event that is not part of the dataFile', function(done) {
            mParticle.forwarder.process({
                EventName: 'eCommerce - AddToCart',
                CustomFlags: {},
                EventDataType: MessageType.Commerce,
                EventCategory: CommerceEventType.AddToCart,
                ProductAction: {
                    ProductActionType: ProductActionType.AddToCart,
                    ProductList: [
                        {
                            Sku: '12345',
                            Category: 'Phones',
                            Brand: 'iPhone 6',
                            Variant: '6',
                            Price: 400,
                            TotalAmount: 400,
                            CouponCode: 'coupon-code',
                            Quantity: 1,
                        },
                    ],
                    TransactionId: 123,
                    Affiliation: 'my-affiliation',
                    TotalAmount: 450,
                    TaxAmount: 40,
                    ShippingAmount: 10,
                    CouponCode: null,
                },
            });

            window.fullStackEventQueue.length.should.equal(0);

            done();
        });
```

it fails because `window['optimizelyClientInstance'].track(eventKey, userId, userAttributes, eventTags);` is always getting called since it is outside the if statement

Reply:
All updates made and additional commerce test added! Take a look and let me know if you spot anything else.